core/arch/riscv/kernel/thread_rv.S

```diff
@@ -0,0 +1,372 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2022-2023 NXP
+ */
+
+#include <asm.S>
+#include <generated/asm-defines.h>
+#include <keep.h>
+#include <kernel/thread.h>
+#include <kernel/thread_private.h>
+#include <mm/core_mmu.h>
+#include <riscv.h>
+#include <riscv_macros.S>
+
+.macro get_thread_ctx res, tmp0
+	lw	\tmp0, THREAD_CORE_LOCAL_CURR_THREAD(tp)
+	la	\res, threads
+1:
+	beqz	\tmp0, 2f
+	addi	\res, \res, THREAD_CTX_SIZE
+	addi	\tmp0, \tmp0, -1
+	bnez	\tmp0, 1b
+2:
+.endm
+
+.macro save_regs, mode
+	addi	sp, sp, -THREAD_TRAP_REGS_SIZE
+.if \mode == TRAP_MODE_USER
+
+	/* Save user thread pointer and load kernel thread pointer */
+	store_xregs sp, THREAD_TRAP_REG_TP, 4, 4
+	load_xregs sp, (THREAD_TRAP_REGS_SIZE - RISCV_XLEN_BYTES), 4, 4
+
+	store_xregs sp, THREAD_TRAP_REG_GP, 3, 3
+
+	/*
+	 * Set the scratch register to 0 such in case of a recursive
+	 * exception thread_trap_vect() knows that it is emitted from kernel.
+	 */
+	csrrw	gp, CSR_XSCRATCH, zero
+	store_xregs sp, THREAD_TRAP_REG_SP, 3, 3
+.option push
+.option norelax
+	la	gp, __global_pointer$
+.option pop
+.endif
+	store_xregs sp, THREAD_TRAP_REG_T3, 28, 31
+	store_xregs sp, THREAD_TRAP_REG_T0, 5, 7
+	store_xregs sp, THREAD_TRAP_REG_A0, 10, 17
+	store_xregs sp, THREAD_TRAP_REG_RA, 1, 1
+
+	csrr	t0, CSR_XSTATUS
+	store_xregs sp, THREAD_TRAP_REG_STATUS, 5, 5
+
+	csrr	a0, CSR_XCAUSE
```

Comment:
Why is this done in this macro? It looks a bit out of place.

Reply:
You mean storing the status when a trap happens ?

Reply:
Yes, when it's done inside a macro called `save_regs()` which saves some general purpose on the stack.